src/main/java/io/vertx/core/http/impl/Http2ServerStream.java

```diff
@@ -60,26 +62,35 @@ class Http2ServerStream extends VertxHttp2Stream<Http2ServerConnection> {
     this.headers = null;
     this.method = method;
     this.uri = uri;
+    this.scheme = null;
     this.host = null;
+    this.authorityString = null;
     this.authority = null;
     this.tracingPolicy = tracingPolicy;
     this.halfClosedRemote = halfClosedRemote;
   }
 
-  Http2ServerStream(Http2ServerConnection conn, ContextInternal context, Http2Headers headers, String serverOrigin, TracingPolicy tracingPolicy, boolean halfClosedRemote) {
+  Http2ServerStream(Http2ServerConnection conn, ContextInternal context, Http2Headers headers, TracingPolicy tracingPolicy, boolean halfClosedRemote) {
     super(conn, context);
 
-    String host = headers.get(":authority") != null ? headers.get(":authority").toString() : null;
-    if (host == null) {
-      int idx = serverOrigin.indexOf("://");
-      host = serverOrigin.substring(idx + 3);
+    CharSequence schemeHeader = headers.getAndRemove(":scheme");
+    HostAndPort authority = null;
+    String authorityString = null;
+    CharSequence authorityHeader = headers.getAndRemove(":authority");
+    if (authorityHeader != null) {
+      authorityString = authorityHeader.toString();
+      authority = HostAndPortImpl.parseHostAndPort(authorityString, -1);
     }
+    CharSequence pathHeader = headers.getAndRemove(":path");
```

Comment:
As said at https://github.com/eclipse-vertx/vert.x/pull/5038#discussion_r1431325625 for the others pseudoheaders